### PR TITLE
Added new test case for Linux VM hibernation with storage workload

### PR DESCRIPTION
**Pull request by juhlee-microsoft**
1. Added a new test case PERF-STORAGE-HIBERNATE
2. Fixed a few lines in POWER-HIBERNATE for VM size adjustment.

**TEST RESULT**
[LISAv2 Test Results Summary]
Test Run On           : 06/10/2020 07:56:21
ARM Image Under Test  : Canonical : UbuntuServer : 16.04-LTS : latest
Initial Kernel Version: 4.15.0-1083-azure
Final Kernel Version  : 5.7.0-next-20200609
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:48

   ID TestArea             TestCaseName                                                      TestResult TestDuration(in minutes) 
---------------------------------------------------------------------------------------------------------------------------------
    1 POWER                POWER-HIBERNATE                                                         PASS                45.99 
***
[LISAv2 Test Results Summary]
Test Run On           : 06/10/2020 07:56:30
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Initial Kernel Version: 5.3.0-1022-azure
Final Kernel Version  : 5.7.0-next-20200609
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:52

   ID TestArea             TestCaseName                                                      TestResult TestDuration(in minutes) 
---------------------------------------------------------------------------------------------------------------------------------
    1 POWER                POWER-HIBERNATE                                                         PASS                50.48 
***
[LISAv2 Test Results Summary]
Test Run On           : 06/10/2020 08:49:46
ARM Image Under Test  : Canonical : UbuntuServer : 16.04-LTS : latest
Initial Kernel Version: 4.15.0-1083-azure
Final Kernel Version  : 5.7.0-next-20200609
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:42

   ID TestArea             TestCaseName                                                      TestResult TestDuration(in minutes) 
---------------------------------------------------------------------------------------------------------------------------------
    1 POWER                POWER-STORAGE-WORKLOAD-HIBERNATE                                        PASS                40.65 
***
[LISAv2 Test Results Summary]
Test Run On           : 06/10/2020 08:49:50
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Initial Kernel Version: 5.3.0-1022-azure
Final Kernel Version  : 5.7.0-next-20200609
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:48

   ID TestArea             TestCaseName                                                      TestResult TestDuration(in minutes) 
---------------------------------------------------------------------------------------------------------------------------------
    1 POWER                POWER-STORAGE-WORKLOAD-HIBERNATE                                        PASS                46.03

### CODE REVIEW

**Comment by dcui** [comment]
I had a quick look and this looks good to me. Thank you, @juhlee-microsoft !

**Review by dcui — COMMENTED** [review]
Looks good to me.

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+	1. Prepare swap space for hibernation
+	2. Compile a new kernel (optional)
+	3. Update the grup.cfg with resume=UUID=xxxx where is from blkid swap disk
+	4. Run the first fio testing
+	5. Hibernate the VM, and verify the VM status
+	5. Resume the VM and verify the VM status.
+	6. Verify no kernel panic or call trace
+	7. Run the second fio testing.
+	8. Verify IOPS counts
+	9. Run the thrid fio testing.
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: third

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+		#region Generate constants.sh
+		# We need to add extra parameters to constants.sh file apart from parameter properties defined in XML.
+		# Hence, we are generating constants.sh file again in test script.
+
+		Write-LogInfo "Generating constants.sh ..."
+		$constantsFile = "$LogDir\constants.sh"
+		foreach ($TestParam in $CurrentTestData.TestParameters.param) {
+			Add-Content -Value "$TestParam" -Path $constantsFile
+			Write-LogInfo "$TestParam added to constants.sh"
+		}
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: I didn't see extra test param here

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+		$ret_val = Update-AzVM -VM $vm -ResourceGroupName $rgName
+		Write-LogInfo "Updated the VM with a new data disk"
+		Write-LogInfo "Waiting for $azureSyncSecond seconds for configuration sync"
+		# Wait for disk sync with Azure host
+		Start-Sleep -s $azureSyncSecond
+
+		# Verify the new data disk addition
+		if ($ret_val.IsSuccessStatusCode) {
+			Write-LogInfo "Successfully add a new disk to the Resource Group, $($rgName)"
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: $rgName is good enough

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+		Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password -command "./SetupHbKernel.sh" -RunInBackground -runAsSudo -ignoreLinuxExitCode:$true | Out-Null
+		Write-LogInfo "Executed SetupHbKernel script inside VM"
+
+		# Wait for kernel compilation completion. 90 min timeout
+		$timeout = New-Timespan -Minutes $maxKernelCompileMin
+		$sw = [diagnostics.stopwatch]::StartNew()
+		while ($sw.elapsed -lt $timeout){
+			$vmCount = $AllVMData.Count
+			Wait-Time -seconds 30
+			$state = Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password "cat /home/$username/state.txt"
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: $username => $user

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+			throw "SetupHbKernel.sh didn't finish in the VM!"
+		}
+
+		# Reboot VM to apply swap setup changes
+		Write-LogInfo "Rebooting All VMs!"
+		$TestProvider.RestartAllDeployments($AllVMData)
+
+		# Check the VM status before hibernation
+		$vmStatus = Get-AzVM -Name $vmName -ResourceGroupName $rgName -Status
+		if ($vmStatus.Statuses[1].DisplayStatus = "VM running") {
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: $vmStatus.Statuses[1].DisplayStatus -eq "VM running"
You need change all on = to -eq, = is for value assignment

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+		# Send fio 32-job command for 5 min
+		Write-LogInfo "Running fio-1 command"
+		$timeout = New-Timespan -Minutes $maxFIORunWaitMin
+		$sw = [diagnostics.stopwatch]::StartNew()
+		$state = Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password -command "bash ./fio1Command.sh" -RunInBackground -runAsSudo
+		Wait-Time -seconds 5
+		while ($sw.elapsed -lt $timeout){
+			Wait-Time -seconds 15
+			$state = Run-LinuxCmd -ip $VMData.PublicIP -port $VMData.SSHPort -username $user -password $password "cat /home/$username/state.txt"
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: ditto

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+				Write-LogInfo "$($state): fio command is still running!"
+			}
+		}
+		if ($state -ne "TestCompleted") {
+			throw "fio-1 command is still running after $maxFIORunWaitMin minutes"
+		}
+
+		# Hibernate the VM
+		Write-LogInfo "Hibernating ..."
+		Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password -command "./test.sh" -runAsSudo -RunInBackground -ignoreLinuxExitCode:$true | Out-Null
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: what is test.sh?

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+		Start-AzVM -Name $vmName -ResourceGroupName $rgName -NoWait | Out-Null
+		Write-LogInfo "Waked up the VM $vmName in Resource Group $rgName and continue checking its status in every 1 minute until $maxVMWakeupMin minutes timeout "
+
+		# Wait for VM resume for $maxVMWakeupMin timeout
+		$timeout = New-Timespan -Minutes $maxVMWakeupMin
+		$sw = [diagnostics.stopwatch]::StartNew()
+		while ($sw.elapsed -lt $timeout){
+			$vmCount = $AllVMData.Count
+			Wait-Time -seconds 60
+			$state = Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password "cat /home/$username/state.txt"
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: ditto for $username

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+		# Send fio 32-job command for 5 min
+		Write-LogInfo "Running fio-2 command"
+		$timeout = New-Timespan -Minutes $maxFIORunWaitMin
+		$sw = [diagnostics.stopwatch]::StartNew()
+		$state = Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password -command "bash ./fio2Command.sh" -RunInBackground -runAsSudo
+		Wait-Time -seconds 5
+		while ($sw.elapsed -lt $timeout){
+			Wait-Time -seconds 15
+			$state = Run-LinuxCmd -ip $VMData.PublicIP -port $VMData.SSHPort -username $user -password $password "cat /home/$username/state.txt"
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: ditto for $username

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+.Description
+	1. Prepare swap space for hibernation
+	2. Compile a new kernel (optional)
+	3. Update the grup.cfg with resume=UUID=xxxx where is from blkid swap disk
+	4. Run the first fio testing
+	5. Hibernate the VM, and verify the VM status
+	5. Resume the VM and verify the VM status.
+	6. Verify no kernel panic or call trace
+	7. Run the second fio testing.
+	8. Verify IOPS counts
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: you didn't verify IOPS in ps1 script

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+		$fio2Command = @"
+source utils.sh
+SetTestStateRunning
+fio --size=1G --name=afterhb --direct=1 --ioengine=libaio --filename=fiodata --overwrite=1 --readwrite=readwrite --bs=1M --runtime=1 --iodepth=128 --numjobs=32 --runtime=300 --output-format=json+ --output=afterhb.json
+rm -f fiodata
+sync
+echo 3 > /proc/sys/vm/drop_caches
+SetTestStateCompleted
+"@
+		Set-Content "$LogDir\fio2Command.sh" $fio2Command
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: you can just use one script, for change part "beforehb", "afterhb", please pass as parameter when invoke the script

**Comment by LiliDeng** [comment]
@juhlee-microsoft I suggest you use the same ps1 script for functional and perf testing, most of the logic is the same. difference is that if it is perf testing, then you add fio verification before and after hibernate, you can use test name or a testparam to distinguish and then decide to run the fio code or not. Also I don't think it is a perf testing, you just run it and didn't collect any results to display the perf data.

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 1 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Fixed in 33b17fcea6c691227d479589631fdd6b9db8a42a

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 2 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: It does. hb_url and hb_branch

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 3 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Fixed in 33b17fcea6c691227d479589631fdd6b9db8a42a

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 4 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Fixed in 33b17fcea6c691227d479589631fdd6b9db8a42a

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 5 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Fixed in 33b17fcea6c691227d479589631fdd6b9db8a42a

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 6 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Fixed in 33b17fcea6c691227d479589631fdd6b9db8a42a

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 7 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: It is generated in SetupHbKernel.sh. hibernation command. This script generation is independent of kernel compilation.

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 8 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Fixed in 33b17fcea6c691227d479589631fdd6b9db8a42a

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 9 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Fixed in 33b17fcea6c691227d479589631fdd6b9db8a42a

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 11 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: It needs to extend to network workload and stress test.  Should I use the single PS script? I think the complexity is high. I won't maintain it then. You still think we should use a single PS script for the existing tests and additional tests?

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 7 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Fixed differently in 1fbc3c0e4f301ee22518072fd3e2d1d4bd93a1ed. Move the test.sh generation to PS1 script.

File: Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1 (hunk 10 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: IOPS or other measurements can be implemented in the next PR. I remove that comment for now.

Comment (juhlee-microsoft) [reply Testscripts/Windows/PERF-STORAGE-FIO-HB.ps1]: Will update later. 5f02ff780c9a63ca11f5fe9c2b6aaa6be88ef158

**Comment by juhlee-microsoft** [comment]
@LiliDeng I fixed all comments and update the test result with the latest one. Please sign-off.

**Review by LiliDeng — DISMISSED** [review]
https://github.com/LIS/LISAv2/blob/master/Testscripts/Windows/Power-Hibernate.ps1#L127 need change.
And squash all commits into one when do merge.

**Review by LiliDeng — APPROVED** [review]

**Pull request merged**